File: src/main/java/com/flowcommerce/core/support/response/ApiResponse.java
package com.flowcommerce.core.support.response;

import com.flowcommerce.core.support.error.ErrorMessage;
import com.flowcommerce.core.support.error.ErrorType;
import lombok.Getter;

@Getter
public class ApiResponse<T> {
    private final ResultType resultType;
    private final T data;
    private final ErrorMessage errorMessage;

    private ApiResponse(ResultType resultType, T data, ErrorMessage errorMessage) {
        this.resultType = resultType;
        this.data = data;
        this.errorMessage = errorMessage;
    }
    public static ApiResponse<Void> success() {
        return new ApiResponse<>(ResultType.SUCCESS, null, null);
    }

    public static <S> ApiResponse<S> success(S data) {
        return new ApiResponse<>(ResultType.SUCCESS, data, null);
    }

    public static <S> ApiResponse<S> error(ErrorType errorType, Object errorData) {
        return new ApiResponse<>(ResultType.ERROR, null, new ErrorMessage(errorType, errorData));
    }

    public static <S> ApiResponse<S> error(ErrorType errorType) {
        return new ApiResponse<>(ResultType.ERROR, null, new ErrorMessage(errorType, null));
    }
}

File: src/main/kotlin/com/flowcommerce/core/support/response/ApiResponse.kt
package com.flowcommerce.core.support.response

import com.flowcommerce.core.support.error.ErrorMessage
import com.flowcommerce.core.support.error.ErrorType

data class ApiResponse<T>(
    val resultType: ResultType,
    val data: T?,
    val errorMessage: ErrorMessage<*>?
) {
    companion object {
        fun success(): ApiResponse<Void> =
            ApiResponse(ResultType.SUCCESS, null, null)

        fun <S> success(data: S): ApiResponse<S> =
            ApiResponse(ResultType.SUCCESS, data, null)

        fun <S> error(errorType: ErrorType, errorData: Any? = null): ApiResponse<S> =
            ApiResponse(ResultType.ERROR, null, ErrorMessage(errorType, errorData))
    }
}
